// webpack-plugin/index.js
// =============================================================================
// Less Setup for Vue
// =============================================================================

// Version:         1.0.1

// Description:     This is a webpack plugin for the less 

// Note:            You will need to install this local module dependencies
//                  - goto less/core/node_modules/webpack-plugin/
//                  - npm install

const path = require('path');
const lessPluginLists = require('less-plugin-lists');
const LessFramework = require('../less-framework.js');
const merge = Object.assign;

const defaults = {
  baseDir: '',
  rootRequire: require,
  id: "styles",
  filename: "../css/styles.css",
  consoleLogs: true,
  consoleLogsDev: false,
  formatLogs: true,
  formatLogsSimplePaths: true
};


// Webpack Plugin (Builds Dependencies)
// =============================================================================

class WebpackPlugin {
  constructor(options) {
    this.options = merge({}, defaults, options);
  }
  apply(compiler) {
    const framework = LessFramework(this.options);
    if (framework.checkPaths()) {
      framework.modulesInit();
    }
  }
  lessLoaderOptions(customOptions) {
    const configuration = {
      paths: [
        this.pathSite,
        this.pathCore
      ],
      plugins: [
        new lessPluginLists()
      ],
      globalVars: {
        "wysiwyg-stylesheet": false,
        "pathSite" : `"${ this.pathSite }"`,
        "pathCore" : `"${ this.pathCore }"`
      }
    };
    return merge({}, configuration, customOptions || null);
  }
}


exports.plugin = WebpackPlugin;

// Less Loader Options helper
// =============================================================================

// const merge = Object.assign();

// exports.paths = {
//   pathCore: path.resolve(__dirname, '../../'),
//   pathSite: path.resolve(__dirname, '../../../site/'),
//   pathJavascript: path.resolve(__dirname, '../../../../js/less-includes.js')
// };
// exports.lessLoaderOptions = {
//   paths: [
//     exports.paths.pathSite,
//     exports.paths.pathCore
//   ],
//   plugins: [
//     new lessPluginLists()
//   ],
//   globalVars: {
//     "wysiwyg-stylesheet": false,
//     "pathSite" : `"${ exports.paths.pathSite }"`,
//     "pathCore" : `"${ exports.paths.pathCore }"`
//   }
// };



// exports.run = function(options) {
  // const framework = LessFramework(merge({},defaults, options));
  // if (framework.checkPaths()) framework.modulesInit();
// };
